service/frontend/domainArchivalConfigStateMachine.go

```diff
@@ -21,216 +21,234 @@
 package frontend
 
 import (
-	"fmt"
 	"github.com/uber/cadence/.gen/go/shared"
 )
 
 type (
-	archivalConfigState struct {
+	// archivalState represents the state of archival config
+	// the only invalid state is {bucket="", status=enabled}
+	// once bucket is set it is immutable
+	// the initial state is {bucket="", status=disabled}, this is what all domains initially default to
+	archivalState struct {
 		bucket string
 		status shared.ArchivalStatus
 	}
 
-	archivalConfigUpdate struct {
-		defaultBucket   string
-		requestedBucket *string
-		requestedStatus *shared.ArchivalStatus
+	// archivalEvent represents a change request to archival config state
+	// the only restriction placed on events is that defaultBucket is not empty
+	// setting requestedBucket to empty string means user is not attempting to set bucket
+	// status can be nil, enabled, or disabled (nil indicates no update by user is being attempted)
+	archivalEvent struct {
+		defaultBucket string
+		bucket        string
+		status        *shared.ArchivalStatus
 	}
 )
 
+// the following errors represents impossible code states that should never occur
 var (
-	errCannotProvideBucket       = &shared.BadRequestError{Message: "Request can only provide bucket name if request also gives status of enabled or disabled."}
-	errDisallowedBucketMetadata  = &shared.BadRequestError{Message: "Cannot set bucket owner or bucket retention (must update bucket manually)."}
-	errStateChangeToNeverEnabled = &shared.BadRequestError{Message: "Cannot change state to never enabled."}
-	errBucketNameUpdate          = &shared.BadRequestError{Message: "Cannot update bucket name after archival status has been moved out of never_enabled for the first time."}
-	errCannotProvideEmptyBucket  = &shared.BadRequestError{Message: "Cannot provide empty bucket name."}
-	errNoDefaultBucket           = &shared.BadRequestError{Message: "No default bucket provided, is cluster correctly configured for archival?"}
-	errCannotHandleStateChange   = &shared.BadRequestError{Message: "Encountered current state and update that cannot be handled (this should be impossible)"}
+	errInvalidState            = &shared.BadRequestError{Message: "Encountered illegal state: archival is enabled but bucket is not set (should be impossible)"}
```

Comment:
It's weird for this class to be returning errors of type: BadRequestErrors. Those errors are for the top level rpc request handlers. This class ideally should be returning just errors.New("..") - and let the upper layer encapsulate into whatever error that makes sense according to caller context. 

On the other hand, if there is an assumption here that the methods in this file are only called within rpc request handlers - just add a comment here

Reply:
Yes logically this class is just a module used by the RPC layer. I will comment it as such.